File: src/runtime/lexer.js
const KEYWORDS = new Set([
  'let', 'func', 'async', 'return', 'if', 'else', 'while', 'for', 'in', 'true', 'false', 'null', 'import', 'as', 'throw', 'try', 'catch', 'component', 'await', 'and', 'or', 'not', 'end'
]);

class Lexer {
  constructor(input, file = '<memory>') {
    this.input = input.replace(/\r\n/g, '\n');
    this.file = file;
    this.pos = 0;
    this.line = 1;
    this.col = 1;
    this.tokens = [];
  }

  tokenize() {
    while (!this.eof()) {
      const ch = this.peek();
      if (ch === ' ' || ch === '\t') {
        this.advance();
        continue;
      }
      if (ch === '\n') {
        this.push('NEWLINE', '\n');
        this.advanceLine();
        continue;
      }
      if (ch === '#') {
        while (!this.eof() && this.peek() !== '\n') this.advance();
        continue;
      }
      if (/[0-9]/.test(ch)) {
        this.tokenizeNumber();
        continue;
      }
      if (/[A-Za-z_]/.test(ch)) {
        this.tokenizeIdentifier();
        continue;
      }
      if (ch === '"' || ch === "'") {
        this.tokenizeString(ch);
        continue;
      }
      this.tokenizeSymbol();
    }
    this.tokens.push(this.mk('EOF', null));
    return this.tokens;
  }

  tokenizeNumber() {
    const start = this.loc();
    let text = '';
    while (!this.eof() && /[0-9.]/.test(this.peek())) text += this.advance();
    this.tokens.push({ type: 'NUMBER', value: Number(text), raw: text, loc: start });
  }

  tokenizeIdentifier() {
    const start = this.loc();
    let text = '';
    while (!this.eof() && /[A-Za-z0-9_]/.test(this.peek())) text += this.advance();
    if (KEYWORDS.has(text)) {
      this.tokens.push({ type: text.toUpperCase(), value: text, loc: start });
    } else {
      this.tokens.push({ type: 'IDENT', value: text, loc: start });
    }
  }

  tokenizeString(quote) {
    const start = this.loc();
    this.advance();
    let out = '';
    while (!this.eof() && this.peek() !== quote) {
      if (this.peek() === '\\') {
        this.advance();
        const esc = this.advance();
        const map = { n: '\n', t: '\t', r: '\r', '"': '"', "'": "'", '\\': '\\' };
        out += map[esc] ?? esc;
      } else {
        out += this.advance();
      }
    }
    this.advance();
    this.tokens.push({ type: 'STRING', value: out, loc: start });
  }

  tokenizeSymbol() {
    const start = this.loc();
    const two = this.input.slice(this.pos, this.pos + 2);
    const pairs = ['==', '!=', '<=', '>=', '=>'];
    if (pairs.includes(two)) {
      this.advance(); this.advance();
      this.tokens.push({ type: two, value: two, loc: start });
      return;
    }
    const one = this.advance();
    const singles = '(){}[],:.+-*/%<>=.';
    if (singles.includes(one)) {
      this.tokens.push({ type: one, value: one, loc: start });
      return;
    }
    throw new Error(`Unexpected token ${one} at ${this.file}:${start.line}:${start.col}`);
  }

  push(type, value) { this.tokens.push(this.mk(type, value)); }
  mk(type, value) { return { type, value, loc: this.loc() }; }
  loc() { return { file: this.file, line: this.line, col: this.col }; }
  eof() { return this.pos >= this.input.length; }
  peek() { return this.input[this.pos]; }
  advance() { const c = this.input[this.pos++]; this.col += 1; return c; }
  advanceLine() { this.pos += 1; this.line += 1; this.col = 1; }
}

module.exports = { Lexer };

File: internal/airt/lexer/lexer.go
package lexer

import "unicode"

type Token struct{ Type, Lit string }

type Lexer struct {
	src []rune
	pos int
}

var kw = map[string]string{"let": "LET", "func": "FUNC", "async": "ASYNC", "return": "RETURN", "if": "IF", "else": "ELSE", "while": "WHILE", "true": "TRUE", "false": "FALSE", "null": "NULL", "import": "IMPORT", "as": "AS", "throw": "THROW", "try": "TRY", "catch": "CATCH", "component": "COMPONENT", "await": "AWAIT", "and": "AND", "or": "OR", "not": "NOT", "end": "END"}

func New(s string) *Lexer  { return &Lexer{src: []rune(s)} }
func (l *Lexer) eof() bool { return l.pos >= len(l.src) }
func (l *Lexer) peek() rune {
	if l.eof() {
		return 0
	}
	return l.src[l.pos]
}
func (l *Lexer) next() rune { r := l.peek(); l.pos++; return r }

func (l *Lexer) Tokens() []Token {
	out := []Token{}
	for !l.eof() {
		c := l.peek()
		switch {
		case c == ' ' || c == '\t' || c == '\r':
			l.next()
		case c == '\n':
			l.next()
			out = append(out, Token{"NEWLINE", "\n"})
		case c == '#':
			for !l.eof() && l.peek() != '\n' {
				l.next()
			}
		case unicode.IsDigit(c):
			out = append(out, l.number())
		case unicode.IsLetter(c) || c == '_':
			out = append(out, l.ident())
		case c == '"' || c == '\'':
			out = append(out, l.str())
		default:
			out = append(out, l.sym())
		}
	}
	return append(out, Token{"EOF", ""})
}

func (l *Lexer) number() Token {
	s := l.pos
	for !l.eof() && (unicode.IsDigit(l.peek()) || l.peek() == '.') {
		l.next()
	}
	return Token{"NUMBER", string(l.src[s:l.pos])}
}
func (l *Lexer) ident() Token {
	s := l.pos
	for !l.eof() && (unicode.IsLetter(l.peek()) || unicode.IsDigit(l.peek()) || l.peek() == '_') {
		l.next()
	}
	t := string(l.src[s:l.pos])
	if k, ok := kw[t]; ok {
		return Token{k, t}
	}
	return Token{"IDENT", t}
}
func (l *Lexer) str() Token {
	q := l.next()
	s := []rune{}
	for !l.eof() && l.peek() != q {
		s = append(s, l.next())
	}
	if !l.eof() {
		l.next()
	}
	return Token{"STRING", string(s)}
}
func (l *Lexer) sym() Token {
	two := ""
	if l.pos+1 < len(l.src) {
		two = string(l.src[l.pos : l.pos+2])
	}
	for _, p := range []string{"==", "!=", "<=", ">="} {
		if two == p {
			l.pos += 2
			return Token{p, p}
		}
	}
	c := string(l.next())
	return Token{c, c}
}
